apex/contrib/groupbn/batch_norm.py

```diff
@@ -102,13 +116,29 @@ def __init__(self, num_features, fuse_relu=False, bn_group=1, max_cta_per_sm=2,
         self.my_data = None
         self.pair_data = None
         self.pair_data2 = None
+        self.pair_data3 = None
         self.local_rank = 0
         self.magic = torch.IntTensor([0])
 
+        #calculate cta per sm occupancies
         assert(max_cta_per_sm>0) # won't be able to do much with 0 CTAs :)
+        self.fwd_occupancy =  min(bnp.bn_fwd_nhwc_occupancy(), max_cta_per_sm)
+        self.bwd_occupancy =  min(bnp.bn_bwd_nhwc_occupancy(), max_cta_per_sm)
+        self.addrelu_fwd_occupancy =  min(bnp.bn_addrelu_fwd_nhwc_occupancy(), max_cta_per_sm)
+        self.addrelu_bwd_occupancy =  min(bnp.bn_addrelu_bwd_nhwc_occupancy(), max_cta_per_sm)
+
+        #calculate grid dimentions based on occupancy numbers
+        mp_count = torch.cuda.get_device_properties(None).multi_processor_count
+        self.fwd_grid_dim_x = max(mp_count*self.fwd_occupancy - cta_launch_margin , 1)
+        self.bwd_grid_dim_x = max(mp_count*self.bwd_occupancy - cta_launch_margin , 1)
+        self.addrelu_fwd_grid_dim_x = max(mp_count*self.addrelu_fwd_occupancy - cta_launch_margin , 1)
+        self.addrelu_bwd_grid_dim_x = max(mp_count*self.addrelu_bwd_occupancy - cta_launch_margin , 1)
+        self.grid_dim_y = (num_features + 63) // 64
+
+        # allocate scratch space used by implementation
+        self.ret_cta = torch.cuda.ByteTensor(8192).fill_(0)
```

Comment:
As @slayton58 kindly pointed out, we don't need to expose our scratch space.
We can instead do
> `auto retired_ctas = at::zeros({shape}, ...)`
> and let the caching allocator do its job.

Reply:
This would require zeroing out this buffer on every call which will come at cost and will be redundant. Functionally it is sufficient to zero the buffer only before the fist use, but then we have to expose it as done here to hold the pointer for reuse.

Reply:
Now I remember that part.
I'm OK with the trade off between perf / cleanliness. What do you think @slayton58 ?

Reply:
I don't think it's friendly, and is exposing complexity to the front-end when I think it could be solved in other ways, but it's your code :)

Reply:
I definitely agree that it's not clean to expose it here.
Considering the purpose of this module, I think it's OK to keep it as is for now. Let's leave a comment here just so we have more context for future discussion. @ekrimer